src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -171,58 +171,71 @@ internal BoundExpression ConvertPatternExpression(TypeSymbol inputType, CSharpSy
-                            LanguageVersion requiredVersion = MessageID.IDS_FeatureGenericPatternMatching.RequiredVersion();
-                            if (requiredVersion > Compilation.LanguageVersion)
-                            {
-                                Error(diagnostics, ErrorCode.ERR_PatternWrongGenericTypeInVersion, typeSyntax,
-                                    operandType, patternType,
-                                    Compilation.LanguageVersion.ToDisplayString(),
-                                    new CSharpRequiredLanguageVersion(requiredVersion));
-                                return true;
-                            }
-
-                            // permit pattern-matching when one of the types is an open type in C# 7.1.
-                            break;
-                        }
-                        else
-                        {
-                            Error(diagnostics, ErrorCode.ERR_PatternWrongType, typeSyntax, operandType, patternType);
+                            Error(diagnostics, ErrorCode.ERR_PatternWrongGenericTypeInVersion, typeSyntax,
+                                operandType, patternType,
+                                Compilation.LanguageVersion.ToDisplayString(),
+                                new CSharpRequiredLanguageVersion(requiredVersion));
                             return true;
                         }
+                    }
+                }
+                else
+                {
+                    Error(diagnostics, ErrorCode.ERR_PatternWrongType, typeSyntax, operandType, patternType);
+                    return true;
                 }
             }
 
             return false;
         }
 
+        /// <summary>
+        /// Does an expression of type <paramref name="expressionType"/> "match" a pattern that looks for
+        /// type <paramref name="patternType"/>?
+        /// 'true' if the matched type catches all of them, 'false' if it catches none of them, and
+        /// 'null' if it might catch some of them.
+        /// </summary>
+        internal static bool? ExpressionOfTypeMatchesPatternType(
+            Conversions conversions,
+            TypeSymbol expressionType,
+            TypeSymbol patternType,
+            ref HashSet<DiagnosticInfo> useSiteDiagnostics,
+            out Conversion conversion,
+            ConstantValue operandConstantValue = null,
+            bool operandCouldBeNull = false)
+        {
+            Debug.Assert((object)expressionType != null);
+            if (expressionType.TypeKind == TypeKind.Dynamic)
+            {
+                // if operand is the dynamic type, we do the same thing as though it were object
+                expressionType = conversions.CorLibrary.GetSpecialType(SpecialType.System_Object);
+            }
+
+            if ((object)expressionType == (object)patternType)
+            {
+                conversion = Conversion.Identity;
+                return true;
```

Comment:
@gafter I see that you deleted the code. Could you please confirm that you added a test that would  fail without this fix?